File: companion-ios/CompanionApp/Integration/BLE/BleRouteSyncPacket.swift
import Foundation

enum BleRouteSyncGattContract {
    static let serviceUUID = "8d0f3f30-7b4d-4f7c-8b24-2f8e7e4e1001"
    static let chunkWriteCharacteristicUUID = "8d0f3f30-7b4d-4f7c-8b24-2f8e7e4e1002"
    static let eventNotifyCharacteristicUUID = "8d0f3f30-7b4d-4f7c-8b24-2f8e7e4e1003"
}

struct RouteTransferChunkEnvelope: Equatable {
    var transferIdentifier: String
    var chunkIndex: Int
    var totalChunks: Int
    var checksumHex: String
    var payloadFragment: Data
}

enum BleRouteSyncPacket: Equatable {
    case chunk(RouteTransferChunkEnvelope)
    case syncMessage(RouteSyncMessage)
}

enum BleRouteSyncPacketError: Error, Equatable {
    case missingHeaderSeparator
    case invalidUtf8Header
    case missingField(String)
    case invalidField(field: String, value: String)
    case payloadLengthMismatch(expected: Int, actual: Int)
    case unsupportedPacketType(String)
    case unsupportedVersion(String)
    case invalidSyncMessage(String)
}

enum BleRouteSyncCodec {
    private static let packetVersion = "1"

    static func canonicalPayloadData(for message: RouteSyncMessage) -> Data {
        Data(canonicalPayloadString(for: message).utf8)
    }

    static func checksumHex(for data: Data) -> String {
        var hash: UInt32 = 2_166_136_261
        for byte in data {
            hash ^= UInt32(byte)
            hash &*= 16_777_619
        }
        return String(format: "%08x", hash)
    }

    static func chunkEnvelopes(
        for message: RouteSyncMessage,
        transferIdentifier: String,
        chunkSizeBytes: Int
    ) -> [RouteTransferChunkEnvelope] {
        let payload = canonicalPayloadData(for: message)
        let safeChunkSize = max(chunkSizeBytes, 1)
        let checksum = checksumHex(for: payload)
        let totalChunks = max(Int(ceil(Double(payload.count) / Double(safeChunkSize))), 1)
        return stride(from: 0, to: payload.count, by: safeChunkSize).enumerated().map { index, offset in
            let end = min(offset + safeChunkSize, payload.count)
            return RouteTransferChunkEnvelope(
                transferIdentifier: transferIdentifier,
                chunkIndex: index,
                totalChunks: totalChunks,
                checksumHex: checksum,
                payloadFragment: payload.subdata(in: offset..<end)
            )
        }
    }

    static func encode(_ packet: BleRouteSyncPacket) -> Data {
        switch packet {
        case .chunk(let chunk):
            return encodePacket(
                headers: [
                    ("v", packetVersion),
                    ("type", "chunk"),
                    ("transfer_id", chunk.transferIdentifier),
                    ("chunk_index", String(chunk.chunkIndex)),
                    ("total_chunks", String(chunk.totalChunks)),
                    ("checksum", chunk.checksumHex),
                ],
                body: chunk.payloadFragment
            )
        case .syncMessage(let message):
            return encodePacket(
                headers: [
                    ("v", packetVersion),
                    ("type", "sync_message"),
                ],
                body: canonicalPayloadData(for: message)
            )
        }
    }

    static func decode(_ payload: Data) throws -> BleRouteSyncPacket {
        let (headers, body) = try splitHeaderAndBody(payload)
        let version = try requiredField("v", in: headers)
        guard version == packetVersion else {
            throw BleRouteSyncPacketError.unsupportedVersion(version)
        }

        switch try requiredField("type", in: headers) {
        case "chunk":
            return .chunk(
                RouteTransferChunkEnvelope(
                    transferIdentifier: try requiredField("transfer_id", in: headers),
                    chunkIndex: try parseInt(try requiredField("chunk_index", in: headers), field: "chunk_index"),
                    totalChunks: try parseInt(try requiredField("total_chunks", in: headers), field: "total_chunks"),
                    checksumHex: try requiredField("checksum", in: headers),
                    payloadFragment: body
                )
            )
        case "sync_message":
            return .syncMessage(try decodeSyncMessage(body))
        case let other:
            throw BleRouteSyncPacketError.unsupportedPacketType(other)
        }
    }

    static func decodeSyncMessage(_ payload: Data) throws -> RouteSyncMessage {
        guard let text = String(data: payload, encoding: .utf8) else {
            throw BleRouteSyncPacketError.invalidSyncMessage("payload was not valid UTF-8")
        }
        return try parseSyncMessage(text)
    }

    private static func canonicalPayloadString(for message: RouteSyncMessage) -> String {
        switch message {
        case .set(let message):
            return routePayloadString(kind: "set", route: message.route)
        case .update(let message):
            return routePayloadString(kind: "update", route: message.route)
        case .clear(let message):
            return [
                "kind=clear",
                "route_id=\(message.routeIdentifier ?? "current")"
            ].joined(separator: "\n")
        case .status(let message):
            return [
                "kind=status",
                "route_id=\(message.routeIdentifier ?? "none")",
                "revision=\(message.revision.map(String.init) ?? "none")",
                "status=\(message.status.rawValue)",
                "detail=\(message.detail ?? "")"
            ].joined(separator: "\n")
        case .rerouteRequest(let message):
            return [
                "kind=reroute_request",
                "route_id=\(message.routeIdentifier)",
                String(format: "rider=%.6f,%.6f", message.riderLocation.latitude, message.riderLocation.longitude),
                "reason=\(message.reason)"
            ].joined(separator: "\n")
        }
    }

    private static func routePayloadString(kind: String, route: NormalizedRoutePackage) -> String {
        let geometry = route.geometry.map {
            String(format: "%.6f,%.6f", $0.latitude, $0.longitude)
        }.joined(separator: ";")
        let maneuvers = route.maneuvers.map { maneuver in
            [
                maneuver.id,
                maneuver.maneuverType.rawValue,
                String(format: "%.1f", maneuver.distanceFromStartMeters),
                maneuver.distanceToNextMeters.map { String(format: "%.1f", $0) } ?? "",
                String(format: "%.6f,%.6f", maneuver.location.latitude, maneuver.location.longitude),
                maneuver.instructionText ?? ""
            ].joined(separator: "|")
        }.joined(separator: ";")

        return [
            "kind=\(kind)",
            "route_id=\(route.routeIdentifier)",
            "revision=\(route.revision)",
            "version=\(route.version.major).\(route.version.minor)",
            String(
                format: "summary=%.1f|%d|%@|%@",
                route.summary.totalDistanceMeters,
                route.summary.estimatedDurationSeconds,
                route.summary.startLabel ?? "",
                route.summary.destinationLabel ?? ""
            ),
            "geometry=\(geometry)",
            "maneuvers=\(maneuvers)",
            "provenance=\(route.provenance.providerID.rawValue)|\(route.provenance.sourceReference ?? "")|\(route.provenance.generatedAtUnixMs)"
        ].joined(separator: "\n")
    }

    private static func parseSyncMessage(_ payload: String) throws -> RouteSyncMessage {
        let fields = Dictionary(uniqueKeysWithValues: payload.split(separator: "\n").compactMap { line in
            guard let separator = line.firstIndex(of: "=") else { return nil }
            let key = String(line[..<separator]).trimmingCharacters(in: .whitespaces)
            let value = String(line[line.index(after: separator)...])
            return (key, value)
        })
        let kind = try requiredField("kind", in: fields)
        switch kind {
        case "set":
            return .set(RouteSetMessage(route: try parseRoutePackage(fields)))
        case "update":
            let route = try parseRoutePackage(fields)
            return .update(
                RouteUpdateMessage(
                    routeIdentifier: route.routeIdentifier,
                    revision: route.revision,
                    route: route
                )
            )
        case "clear":
            return .clear(RouteClearMessage(routeIdentifier: optionalStringField("route_id", in: fields)))
        case "status":
            return .status(
                RouteStatusMessage(
                    routeIdentifier: optionalStringField("route_id", in: fields),
                    revision: try optionalIntField("revision", in: fields),
                    status: try parseStatus(try requiredField("status", in: fields)),
                    detail: optionalStringField("detail", in: fields)
                )
            )
        case "reroute_request":
            return .rerouteRequest(
                RouteRerouteRequestMessage(
                    routeIdentifier: optionalStringField("route_id", in: fields) ?? "none",
                    riderLocation: try parseCoordinatePoint(try requiredField("rider", in: fields)),
                    reason: try requiredField("reason", in: fields)
                )
            )
        default:
            throw BleRouteSyncPacketError.invalidSyncMessage("Unsupported kind \(kind)")
        }
    }

    private static func parseRoutePackage(_ fields: [String: String]) throws -> NormalizedRoutePackage {
        let version = try requiredField("version", in: fields)
        let versionParts = version.split(separator: ".", maxSplits: 1, omittingEmptySubsequences: false)
        guard versionParts.count == 2 else {
            throw BleRouteSyncPacketError.invalidField(field: "version", value: version)
        }
        let summary = try parseSummary(try requiredField("summary", in: fields))
        let provenance = try parseProvenance(try requiredField("provenance", in: fields))
        return NormalizedRoutePackage(
            version: RoutePackageVersion(
                major: UInt16(try parseInt(String(versionParts[0]), field: "version.major")),
                minor: UInt16(try parseInt(String(versionParts[1]), field: "version.minor"))
            ),
            routeIdentifier: try requiredField("route_id", in: fields),
            revision: try parseInt(try requiredField("revision", in: fields), field: "revision"),
            geometry: try parseGeometry(try requiredField("geometry", in: fields)),
            maneuvers: try parseManeuvers(try requiredField("maneuvers", in: fields)),
            summary: summary,
            provenance: provenance
        )
    }

    private static func parseSummary(_ value: String) throws -> RouteSummary {
        let parts = value.split(separator: "|", maxSplits: 3, omittingEmptySubsequences: false).map(String.init)
        guard parts.count == 4 else {
            throw BleRouteSyncPacketError.invalidField(field: "summary", value: value)
        }
        return RouteSummary(
            totalDistanceMeters: try parseDouble(parts[0], field: "summary.totalDistanceMeters"),
            estimatedDurationSeconds: try parseInt(parts[1], field: "summary.estimatedDurationSeconds"),
            startLabel: nonEmpty(parts[2]),
            destinationLabel: nonEmpty(parts[3])
        )
    }

    private static func parseGeometry(_ value: String) throws -> [CoordinatePoint] {
        guard !value.isEmpty else { return [] }
        return try value.split(separator: ";", omittingEmptySubsequences: false).map {
            try parseCoordinatePoint(String($0))
        }
    }

    private static func parseManeuvers(_ value: String) throws -> [RouteManeuver] {
        guard !value.isEmpty else { return [] }
        return try value.split(separator: ";", omittingEmptySubsequences: false).map { entry in
            let parts = entry.split(separator: "|", maxSplits: 5, omittingEmptySubsequences: false).map(String.init)
            let distanceFromStart = try parseDouble(parts[safe: 2] ?? "", field: "maneuver.distanceFromStartMeters")
            let tuple: (Double?, String, String)
            switch parts.count {
            case 5:
                tuple = (nil, parts[3], parts[4])
            case 6:
                tuple = (
                    nonEmpty(parts[3]).map { try parseDouble($0, field: "maneuver.distanceToNextMeters") },
                    parts[4],
                    parts[5]
                )
            default:
                throw BleRouteSyncPacketError.invalidField(field: "maneuvers", value: String(entry))
            }
            return RouteManeuver(
                id: parts[safe: 0] ?? "",
                maneuverType: try parseManeuverType(parts[safe: 1] ?? ""),
                location: try parseCoordinatePoint(tuple.1),
                distanceFromStartMeters: distanceFromStart,
                distanceToNextMeters: tuple.0,
                instructionText: nonEmpty(tuple.2)
            )
        }
    }

    private static func parseProvenance(_ value: String) throws -> RouteProvenance {
        let parts = value.split(separator: "|", maxSplits: 2, omittingEmptySubsequences: false).map(String.init)
        guard parts.count == 3 else {
            throw BleRouteSyncPacketError.invalidField(field: "provenance", value: value)
        }
        return RouteProvenance(
            providerID: parseProvider(parts[0]),
            sourceReference: nonEmpty(parts[1]),
            generatedAtUnixMs: UInt64(try parseInt(parts[2], field: "provenance.generatedAtUnixMs"))
        )
    }

    private static func parseCoordinatePoint(_ value: String) throws -> CoordinatePoint {
        let parts = value.split(separator: ",", maxSplits: 1, omittingEmptySubsequences: false).map(String.init)
        guard parts.count == 2 else {
            throw BleRouteSyncPacketError.invalidField(field: "coordinate", value: value)
        }
        return CoordinatePoint(
            latitude: try parseDouble(parts[0], field: "coordinate.latitude"),
            longitude: try parseDouble(parts[1], field: "coordinate.longitude")
        )
    }

    private static func parseProvider(_ value: String) -> RouteProviderID {
        switch value.trimmingCharacters(in: .whitespacesAndNewlines) {
        case "hsl", "hsl_digitransit": return .hsl
        case "google_ingest", "googleIngest": return .googleIngest
        case "osm": return .osm
        case "gpx": return .gpxImport
        case "fit": return .fitImport
        case "tcx": return .tcxImport
        case "garmin_api", "garminApi": return .garminApi
        case "garmin_file", "garminFile": return .garminFile
        default: return .osm
        }
    }

    private static func parseManeuverType(_ value: String) throws -> RouteManeuverType {
        switch value.trimmingCharacters(in: .whitespacesAndNewlines) {
        case "depart": return .depart
        case "straight": return .straight
        case "slight_left", "slightLeft": return .slightLeft
        case "left": return .left
        case "sharp_left", "sharpLeft": return .sharpLeft
        case "slight_right", "slightRight": return .slightRight
        case "right": return .right
        case "sharp_right", "sharpRight": return .sharpRight
        case "uturn", "u_turn": return .uturn
        case "roundabout": return .roundabout
        case "merge": return .merge
        case "ramp": return .ramp
        case "arrive": return .arrive
        default:
            throw BleRouteSyncPacketError.invalidField(field: "maneuverType", value: value)
        }
    }

    private static func parseStatus(_ value: String) throws -> RouteSyncStatusCode {
        switch value.trimmingCharacters(in: .whitespacesAndNewlines) {
        case "accepted": return .accepted
        case "applying": return .applying
        case "active": return .active
        case "cleared": return .cleared
        case "rejected": return .rejected
        case "retryable_failure", "retryableFailure": return .retryableFailure
        case "fatal_failure", "fatalFailure": return .fatalFailure
        default:
            throw BleRouteSyncPacketError.invalidField(field: "status", value: value)
        }
    }

    private static func encodePacket(headers: [(String, String)], body: Data) -> Data {
        var encoded = headers.map { "\($0.0)=\($0.1)" }.joined(separator: "\n")
        encoded.append("\npayload_length=\(body.count)\n\n")
        var data = Data(encoded.utf8)
        data.append(body)
        return data
    }

    private static func splitHeaderAndBody(_ payload: Data) throws -> ([String: String], Data) {
        guard let separatorRange = payload.range(of: Data("\n\n".utf8)) else {
            throw BleRouteSyncPacketError.missingHeaderSeparator
        }
        let headerData = payload.subdata(in: 0..<separatorRange.lowerBound)
        let body = payload.subdata(in: separatorRange.upperBound..<payload.count)
        guard let headerText = String(data: headerData, encoding: .utf8) else {
            throw BleRouteSyncPacketError.invalidUtf8Header
        }
        let headers = Dictionary(uniqueKeysWithValues: headerText.split(separator: "\n").compactMap { line in
            guard let separator = line.firstIndex(of: "=") else { return nil }
            let key = String(line[..<separator]).trimmingCharacters(in: .whitespaces)
            let value = String(line[line.index(after: separator)...])
            return (key, value)
        })
        let expectedLength = try parseInt(try requiredField("payload_length", in: headers), field: "payload_length")
        guard expectedLength == body.count else {
            throw BleRouteSyncPacketError.payloadLengthMismatch(expected: expectedLength, actual: body.count)
        }
        return (headers, body)
    }

    private static func requiredField(_ key: String, in headers: [String: String]) throws -> String {
        guard let value = headers[key] else {
            throw BleRouteSyncPacketError.missingField(key)
        }
        return value
    }

    private static func optionalStringField(_ key: String, in headers: [String: String]) -> String? {
        guard let value = headers[key], !value.isEmpty, value != "none", value != "current" else {
            return nil
        }
        return value
    }

    private static func optionalIntField(_ key: String, in headers: [String: String]) throws -> Int? {
        guard let value = headers[key], !value.isEmpty, value != "none" else {
            return nil
        }
        return try parseInt(value, field: key)
    }

    private static func parseInt(_ value: String, field: String) throws -> Int {
        guard let intValue = Int(value) else {
            throw BleRouteSyncPacketError.invalidField(field: field, value: value)
        }
        return intValue
    }

    private static func parseDouble(_ value: String, field: String) throws -> Double {
        guard let doubleValue = Double(value) else {
            throw BleRouteSyncPacketError.invalidField(field: field, value: value)
        }
        return doubleValue
    }

    private static func nonEmpty(_ value: String) -> String? {
        value.isEmpty ? nil : value
    }
}

private extension Collection {
    subscript(safe index: Index) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

File: companion-android/app/src/main/java/me/fiksu/esp32map/companion/integration/ble/BleRouteSyncPacket.kt
package me.fiksu.esp32map.companion.integration.ble

import kotlin.math.ceil
import me.fiksu.esp32map.companion.domain.CoordinatePoint
import me.fiksu.esp32map.companion.domain.NormalizedRoutePackage
import me.fiksu.esp32map.companion.domain.RouteClearMessage
import me.fiksu.esp32map.companion.domain.RouteManeuver
import me.fiksu.esp32map.companion.domain.RouteManeuverType
import me.fiksu.esp32map.companion.domain.RoutePackageVersion
import me.fiksu.esp32map.companion.domain.RouteProvenance
import me.fiksu.esp32map.companion.domain.RouteProviderId
import me.fiksu.esp32map.companion.domain.RouteRerouteRequestMessage
import me.fiksu.esp32map.companion.domain.RouteSetMessage
import me.fiksu.esp32map.companion.domain.RouteStatusMessage
import me.fiksu.esp32map.companion.domain.RouteSyncMessage
import me.fiksu.esp32map.companion.domain.RouteSyncStatusCode
import me.fiksu.esp32map.companion.domain.RouteSummary
import me.fiksu.esp32map.companion.domain.RouteUpdateMessage

object BleRouteSyncGattContract {
    const val SERVICE_UUID: String = "8d0f3f30-7b4d-4f7c-8b24-2f8e7e4e1001"
    const val CHUNK_WRITE_CHARACTERISTIC_UUID: String = "8d0f3f30-7b4d-4f7c-8b24-2f8e7e4e1002"
    const val EVENT_NOTIFY_CHARACTERISTIC_UUID: String = "8d0f3f30-7b4d-4f7c-8b24-2f8e7e4e1003"
}

data class RouteTransferChunkEnvelope(
    val transferIdentifier: String,
    val chunkIndex: Int,
    val totalChunks: Int,
    val checksumHex: String,
    val payloadFragment: ByteArray,
)

sealed interface BleRouteSyncPacket {
    data class Chunk(val chunk: RouteTransferChunkEnvelope) : BleRouteSyncPacket
    data class SyncMessage(val message: RouteSyncMessage) : BleRouteSyncPacket
}

object BleRouteSyncCodec {
    private const val PACKET_VERSION: String = "1"

    fun canonicalPayloadBytes(message: RouteSyncMessage): ByteArray =
        canonicalPayloadString(message).encodeToByteArray()

    fun checksumHex(payload: ByteArray): String {
        var hash = 2_166_136_261L
        payload.forEach { byte ->
            hash = hash xor (byte.toInt() and 0xff).toLong()
            hash = (hash * 16_777_619L) and 0xffff_ffffL
        }
        return "%08x".format(hash)
    }

    fun chunkEnvelopes(
        message: RouteSyncMessage,
        transferIdentifier: String,
        chunkSizeBytes: Int,
    ): List<RouteTransferChunkEnvelope> {
        val payload = canonicalPayloadBytes(message)
        val safeChunkSize = maxOf(chunkSizeBytes, 1)
        val totalChunks = maxOf(1, ceil(payload.size.toDouble() / safeChunkSize.toDouble()).toInt())
        val checksum = checksumHex(payload)
        return payload.asList().chunked(safeChunkSize).mapIndexed { index, fragment ->
            RouteTransferChunkEnvelope(
                transferIdentifier = transferIdentifier,
                chunkIndex = index,
                totalChunks = totalChunks,
                checksumHex = checksum,
                payloadFragment = fragment.toByteArray(),
            )
        }
    }

    fun encode(packet: BleRouteSyncPacket): ByteArray {
        return when (packet) {
            is BleRouteSyncPacket.Chunk -> encodePacket(
                headers = listOf(
                    "v" to PACKET_VERSION,
                    "type" to "chunk",
                    "transfer_id" to packet.chunk.transferIdentifier,
                    "chunk_index" to packet.chunk.chunkIndex.toString(),
                    "total_chunks" to packet.chunk.totalChunks.toString(),
                    "checksum" to packet.chunk.checksumHex,
                ),
                body = packet.chunk.payloadFragment,
            )
            is BleRouteSyncPacket.SyncMessage -> encodePacket(
                headers = listOf(
                    "v" to PACKET_VERSION,
                    "type" to "sync_message",
                ),
                body = canonicalPayloadBytes(packet.message),
            )
        }
    }

    fun decode(payload: ByteArray): BleRouteSyncPacket {
        val (headers, body) = splitHeaderAndBody(payload)
        require(requiredField(headers, "v") == PACKET_VERSION) { "Unsupported packet version" }
        return when (requiredField(headers, "type")) {
            "chunk" -> BleRouteSyncPacket.Chunk(
                RouteTransferChunkEnvelope(
                    transferIdentifier = requiredField(headers, "transfer_id"),
                    chunkIndex = parseInt(requiredField(headers, "chunk_index"), "chunk_index"),
                    totalChunks = parseInt(requiredField(headers, "total_chunks"), "total_chunks"),
                    checksumHex = requiredField(headers, "checksum"),
                    payloadFragment = body,
                ),
            )
            "sync_message" -> BleRouteSyncPacket.SyncMessage(decodeSyncMessage(body))
            else -> error("Unsupported BLE route sync packet type")
        }
    }

    fun decodeSyncMessage(payload: ByteArray): RouteSyncMessage {
        val text = payload.decodeToString()
        val fields = text.lineSequence()
            .mapNotNull { line ->
                val separator = line.indexOf('=')
                if (separator < 0) null else line.substring(0, separator).trim() to line.substring(separator + 1)
            }
            .toMap()
        return when (val kind = requiredField(fields, "kind")) {
            "set" -> RouteSyncMessage.Set(RouteSetMessage(parseRoutePackage(fields)))
            "update" -> {
                val route = parseRoutePackage(fields)
                RouteSyncMessage.Update(
                    RouteUpdateMessage(
                        routeIdentifier = route.routeIdentifier,
                        revision = route.revision,
                        route = route,
                    ),
                )
            }
            "clear" -> RouteSyncMessage.Clear(RouteClearMessage(optionalStringField(fields, "route_id")))
            "status" -> RouteSyncMessage.Status(
                RouteStatusMessage(
                    routeIdentifier = optionalStringField(fields, "route_id"),
                    revision = optionalIntField(fields, "revision"),
                    status = parseStatus(requiredField(fields, "status")),
                    detail = optionalStringField(fields, "detail"),
                ),
            )
            "reroute_request" -> RouteSyncMessage.RerouteRequest(
                RouteRerouteRequestMessage(
                    routeIdentifier = optionalStringField(fields, "route_id") ?: "none",
                    riderLocation = parseCoordinatePoint(requiredField(fields, "rider")),
                    reason = requiredField(fields, "reason"),
                ),
            )
            else -> error("Unsupported sync message kind $kind")
        }
    }

    private fun canonicalPayloadString(message: RouteSyncMessage): String {
        return when (message) {
            is RouteSyncMessage.Set -> routePayloadString("set", message.message.route)
            is RouteSyncMessage.Update -> routePayloadString("update", message.message.route)
            is RouteSyncMessage.Clear -> listOf(
                "kind=clear",
                "route_id=${message.message.routeIdentifier ?: "current"}",
            ).joinToString("\n")
            is RouteSyncMessage.Status -> listOf(
                "kind=status",
                "route_id=${message.message.routeIdentifier ?: "none"}",
                "revision=${message.message.revision ?: "none"}",
                "status=${message.message.status.name.lowercase()}",
                "detail=${message.message.detail.orEmpty()}",
            ).joinToString("\n")
            is RouteSyncMessage.RerouteRequest -> listOf(
                "kind=reroute_request",
                "route_id=${message.message.routeIdentifier}",
                "rider=${"%.6f,%.6f".format(message.message.riderLocation.latitude, message.message.riderLocation.longitude)}",
                "reason=${message.message.reason}",
            ).joinToString("\n")
        }
    }

    private fun routePayloadString(kind: String, route: NormalizedRoutePackage): String {
        val geometry = route.geometry.joinToString(";") { point ->
            "%.6f,%.6f".format(point.latitude, point.longitude)
        }
        val maneuvers = route.maneuvers.joinToString(";") { maneuver ->
            listOf(
                maneuver.id,
                maneuver.maneuverType.name.lowercase(),
                "%.1f".format(maneuver.distanceFromStartMeters),
                maneuver.distanceToNextMeters?.let { "%.1f".format(it) }.orEmpty(),
                "%.6f,%.6f".format(maneuver.location.latitude, maneuver.location.longitude),
                maneuver.instructionText.orEmpty(),
            ).joinToString("|")
        }
        return listOf(
            "kind=$kind",
            "route_id=${route.routeIdentifier}",
            "revision=${route.revision}",
            "version=${route.version.major}.${route.version.minor}",
            "summary=${"%.1f|%d|%s|%s".format(route.summary.totalDistanceMeters, route.summary.estimatedDurationSeconds, route.summary.startLabel.orEmpty(), route.summary.destinationLabel.orEmpty())}",
            "geometry=$geometry",
            "maneuvers=$maneuvers",
            "provenance=${route.provenance.providerId.name.lowercase()}|${route.provenance.sourceReference.orEmpty()}|${route.provenance.generatedAtUnixMs}",
        ).joinToString("\n")
    }

    private fun parseRoutePackage(fields: Map<String, String>): NormalizedRoutePackage {
        val version = requiredField(fields, "version")
        val versionParts = version.split('.', limit = 2)
        require(versionParts.size == 2) { "Invalid version $version" }
        return NormalizedRoutePackage(
            version = RoutePackageVersion(
                major = parseInt(versionParts[0], "version.major"),
                minor = parseInt(versionParts[1], "version.minor"),
            ),
            routeIdentifier = requiredField(fields, "route_id"),
            revision = parseInt(requiredField(fields, "revision"), "revision"),
            geometry = parseGeometry(requiredField(fields, "geometry")),
            maneuvers = parseManeuvers(requiredField(fields, "maneuvers")),
            summary = parseSummary(requiredField(fields, "summary")),
            provenance = parseProvenance(requiredField(fields, "provenance")),
        )
    }

    private fun parseSummary(value: String): RouteSummary {
        val parts = value.split('|', limit = 4)
        require(parts.size == 4) { "Invalid summary $value" }
        return RouteSummary(
            totalDistanceMeters = parseDouble(parts[0], "summary.totalDistanceMeters"),
            estimatedDurationSeconds = parseInt(parts[1], "summary.estimatedDurationSeconds"),
            startLabel = parts[2].ifEmpty { null },
            destinationLabel = parts[3].ifEmpty { null },
        )
    }

    private fun parseGeometry(value: String): List<CoordinatePoint> {
        if (value.isEmpty()) return emptyList()
        return value.split(';').map(::parseCoordinatePoint)
    }

    private fun parseManeuvers(value: String): List<RouteManeuver> {
        if (value.isEmpty()) return emptyList()
        return value.split(';').map { entry ->
            val parts = entry.split('|', limit = 6)
            val distanceFromStart = parseDouble(parts.getOrElse(2) { "" }, "maneuver.distanceFromStartMeters")
            val tuple = when (parts.size) {
                5 -> Triple(null, parts[3], parts[4])
                6 -> Triple(parts[3].ifEmpty { null }?.let { parseDouble(it, "maneuver.distanceToNextMeters") }, parts[4], parts[5])
                else -> error("Invalid maneuver entry $entry")
            }
            RouteManeuver(
                id = parts.firstOrNull().orEmpty(),
                maneuverType = parseManeuverType(parts.getOrElse(1) { "" }),
                location = parseCoordinatePoint(tuple.second),
                distanceFromStartMeters = distanceFromStart,
                distanceToNextMeters = tuple.first,
                instructionText = tuple.third.ifEmpty { null },
            )
        }
    }

    private fun parseProvenance(value: String): RouteProvenance {
        val parts = value.split('|', limit = 3)
        require(parts.size == 3) { "Invalid provenance $value" }
        return RouteProvenance(
            providerId = parseProvider(parts[0]),
            sourceReference = parts[1].ifEmpty { null },
            generatedAtUnixMs = parseLong(parts[2], "provenance.generatedAtUnixMs"),
        )
    }

    private fun parseCoordinatePoint(value: String): CoordinatePoint {
        val parts = value.split(',', limit = 2)
        require(parts.size == 2) { "Invalid coordinate $value" }
        return CoordinatePoint(
            latitude = parseDouble(parts[0], "coordinate.latitude"),
            longitude = parseDouble(parts[1], "coordinate.longitude"),
        )
    }

    private fun parseProvider(value: String): RouteProviderId {
        return when (value.trim()) {
            "hsl", "hsl_digitransit" -> RouteProviderId.HSL
            "osm" -> RouteProviderId.OSM
            "gpx" -> RouteProviderId.GPX_IMPORT
            "fit" -> RouteProviderId.FIT_IMPORT
            "tcx" -> RouteProviderId.TCX_IMPORT
            else -> RouteProviderId.OSM
        }
    }

    private fun parseManeuverType(value: String): RouteManeuverType {
        return when (value.trim()) {
            "depart" -> RouteManeuverType.DEPART
            "straight" -> RouteManeuverType.STRAIGHT
            "slight_left", "slightLeft" -> RouteManeuverType.SLIGHT_LEFT
            "left" -> RouteManeuverType.LEFT
            "sharp_left", "sharpLeft" -> RouteManeuverType.SHARP_LEFT
            "slight_right", "slightRight" -> RouteManeuverType.SLIGHT_RIGHT
            "right" -> RouteManeuverType.RIGHT
            "sharp_right", "sharpRight" -> RouteManeuverType.SHARP_RIGHT
            "uturn", "u_turn" -> RouteManeuverType.UTURN
            "roundabout" -> RouteManeuverType.ROUNDABOUT
            "merge" -> RouteManeuverType.MERGE
            "ramp" -> RouteManeuverType.RAMP
            "arrive" -> RouteManeuverType.ARRIVE
            else -> error("Unsupported maneuver type $value")
        }
    }

    private fun parseStatus(value: String): RouteSyncStatusCode {
        return when (value.trim()) {
            "accepted" -> RouteSyncStatusCode.ACCEPTED
            "applying" -> RouteSyncStatusCode.APPLYING
            "active" -> RouteSyncStatusCode.ACTIVE
            "cleared" -> RouteSyncStatusCode.CLEARED
            "rejected" -> RouteSyncStatusCode.REJECTED
            "retryable_failure", "retryableFailure" -> RouteSyncStatusCode.RETRYABLE_FAILURE
            "fatal_failure", "fatalFailure" -> RouteSyncStatusCode.FATAL_FAILURE
            else -> error("Unsupported route sync status $value")
        }
    }

    private fun encodePacket(headers: List<Pair<String, String>>, body: ByteArray): ByteArray {
        val header = buildString {
            append(headers.joinToString("\n") { (key, value) -> "$key=$value" })
            append("\npayload_length=${body.size}\n\n")
        }
        return header.encodeToByteArray() + body
    }

    private fun splitHeaderAndBody(payload: ByteArray): Pair<Map<String, String>, ByteArray> {
        val separator = payload.indexOfHeaderSeparator()
        require(separator >= 0) { "Missing BLE route sync header separator" }
        val headerText = payload.copyOfRange(0, separator).decodeToString()
        val body = payload.copyOfRange(separator + 2, payload.size)
        val headers = headerText.lineSequence()
            .mapNotNull { line ->
                val index = line.indexOf('=')
                if (index < 0) null else line.substring(0, index).trim() to line.substring(index + 1)
            }
            .toMap()
        val expectedLength = parseInt(requiredField(headers, "payload_length"), "payload_length")
        require(expectedLength == body.size) { "BLE route sync payload length mismatch" }
        return headers to body
    }

    private fun requiredField(fields: Map<String, String>, key: String): String =
        fields[key] ?: error("Missing field $key")

    private fun optionalStringField(fields: Map<String, String>, key: String): String? =
        fields[key]?.takeUnless { it.isEmpty() || it == "none" || it == "current" }

    private fun optionalIntField(fields: Map<String, String>, key: String): Int? =
        fields[key]?.takeUnless { it.isEmpty() || it == "none" }?.let { parseInt(it, key) }

    private fun parseInt(value: String, field: String): Int =
        value.toIntOrNull() ?: error("Invalid integer for $field: $value")

    private fun parseLong(value: String, field: String): Long =
        value.toLongOrNull() ?: error("Invalid long for $field: $value")

    private fun parseDouble(value: String, field: String): Double =
        value.toDoubleOrNull() ?: error("Invalid double for $field: $value")
}

private fun ByteArray.indexOfHeaderSeparator(): Int {
    for (index in 0 until size - 1) {
        if (this[index] == '\n'.code.toByte() && this[index + 1] == '\n'.code.toByte()) {
            return index
        }
    }
    return -1
}
